modules/src/ics04_channel/channel.rs

```diff
@@ -22,30 +22,47 @@ pub struct ChannelEnd {
 impl TryFromRaw for ChannelEnd {
     type RawType = RawChannel;
     type Error = anomaly::Error<Kind>;
+
     fn try_from(value: RawChannel) -> Result<Self, Self::Error> {
-        // Todo: Do validation of data here. This is just an example implementation for testing.
-        let ordering = match value.ordering {
-            0 => Order::None,
-            1 => Order::Unordered,
-            2 => Order::Ordered,
-            _ => panic!("invalid order number"),
+        // Parse the ordering type. Propagate the error, if any, to our caller.
+        let chan_ordering = Order::from_i32(value.ordering)?;
+
+        let chan_state = State::from_i32(value.state)?;
+
+        // Pop out the Counterparty from the Option.
+        let counterparty = match value.counterparty {
+            Some(cp) => cp,
+            None => return Err(Kind::MissingCounterparty.into()),
         };
-        let counterparty = value.counterparty.unwrap();
+
+        // Assemble the 'remote' attribute of the Channel, which represents the Counterparty.
         let remote = Counterparty {
-            port_id: PortId::from_str(counterparty.port_id.as_str()).unwrap(),
-            channel_id: ChannelId::from_str(counterparty.channel_id.as_str()).unwrap(),
+            port_id: PortId::from_str(counterparty.port_id.as_str())
+                .map_err(|e| Kind::IdentifierError.context(e))?,
+            channel_id: ChannelId::from_str(counterparty.channel_id.as_str())
+                .map_err(|e| Kind::IdentifierError.context(e))?,
         };
+
+        // Parse each item in connection_hops into a ConnectionId.
         let connection_hops = value
             .connection_hops
             .into_iter()
-            .map(|e| ConnectionId::from_str(e.as_str()).unwrap())
-            .collect();
+            .map(|conn_id| ConnectionId::from_str(conn_id.as_str()))
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|e| Kind::IdentifierError.context(e))?;
+
+        // No explicit validation is necessary for the version attribute (opaque field).
```

Comment:
Then what's the point of it? :)
Is it checked in some other context?

Reply:
The full details on the version field are here:
https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics#versioning

The version field may be checked in other contexts (e.g., during a handshake, but as far as I can tell, it is not). But the point is that this field may contain fairly arbitrary data, according to the ICS4 spec. Looking through the cosmos-sdk codebase, however, I see that there is some basic validation on this field:

https://github.com/cosmos/cosmos-sdk/blob/e6a56226a88ec4e8761b1388a3a666af627d9e2e/x/ibc/04-channel/types/channel.go#L72

I'm adding a TODO to clarify this.

Reply:
It seems that there might be a mistake in cosmos-sdk code. We should permit an empty version string for ChannelEnd, so basically no validation is required here.

https://github.com/cosmos/cosmos-sdk/issues/6830